src/sage/schemes/projective/coherent_sheaf.py

```diff
@@ -0,0 +1,323 @@
+r"""
+Coherent sheaves on projective schemes
+
+EXAMPLES:
+
+We define the Fermat cubic surface in `\PP^2` and examine its structure sheaf::
+
+    sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+    sage: X = P2.subscheme(x^4 + y^4 + z^4)
+    sage: sh = X.structure_sheaf()
+    sage: sh
+    Coherent sheaf on Closed subscheme of Projective Space of dimension 2
+     over Rational Field defined by: x^4 + y^4 + z^4
+    sage: sh.euler_characteristic()
+    -2
+
+AUTHORS:
+
+- Kwankyu Lee (2024-01-22): initial version
+
+"""
+
+from functools import cached_property
+from sage.structure.sage_object import SageObject
+from sage.modules.free_module import FreeModule
+
+
+class CoherentSheaf(SageObject):
+    r"""
+    Coherent sheaf on a projective scheme.
+
+    INPUT:
+
+    - ``scheme`` -- the base scheme on which the sheaf is defined
+
+    - ``module`` -- a free module or its quotient by a submodule
+
+    - ``twist`` -- (default: 0) an integer
+
+    This class constructs the coherent sheaf `\tilde M(n)` if `M` is the
+    ``module`` and `n` is the ``twist``.
+    """
+    def __init__(self, scheme, module, twist=0):
+        """
+        Initialize ``self``.
+
+        TESTS::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme(x^4 + y^4 + z^4)
+            sage: sheaf = X.structure_sheaf()
+            sage: TestSuite(sheaf).run(skip=['_test_pickling'])
+        """
+        try:
+            if module.is_ambient():
+                module = module.quotient(module.zero_submodule())
+        except AttributeError:
+            pass
+
+        assert module.cover() == module.free_cover()
+
+        self._base_scheme = scheme
+        self._module = module
+        self._twist = twist
+
+    @cached_property
```

Comment:
```suggestion
@abstract_method(optional=True)
```
Since the default is to raise an error, there is no need to cache it and this will be overwritten by subclasses. This would be better for the developer IMO (and you wouldn't need to raise the error).

Reply:
Right. Thanks for catching this.

Removed `@cached_property`.